Multicluster (VPN-mode) stability scenarios

philrud opened the pull request:
Introduces a stability chart with probe apps for various multicluster scenarios
and setup scripts that spin up a corresponding multicluster-aware Istio installation
in a separate (external) cluster using the 'new' Istio installer.

howardjohn commented on perf/stability/multicluster-vpn/templates/default.yaml:
```diff
@@ -0,0 +1,40 @@
+  selector:
+    app: srv-default
+---
+apiVersion: apps/v1beta1
```
nit: you can use apps/v1

howardjohn commented on perf/stability/multicluster-vpn/templates/locality-failover.yaml:
```diff
@@ -0,0 +1,40 @@
+metadata:
+  name: srv-locality-failover
+spec:
+  replicas: 4
```
nit: maybe make configurable in helm?

philrud replied:
I'm planning on kustomize but this is WIP

costinm replied:
+1 on doing it via kustomize - I think we could just have it kustomize first, without any helm template.

howardjohn commented on setup/stability-extra/multicluster-vpn/default/setup.sh:
```diff
@@ -0,0 +1,60 @@
+  -f ${TEMP_DIR}/values-${SCENARIO}.yaml \
+  | $KUBECTL2 apply -f -
+
+helm template \
```
I don't think we need this if we are using manual injection?

philrud replied:
Even for manual injection we need an injector configmap. I found it more straightforward to simply install autoinjector and point istioctl to that NS. Also I would be cool to get the autoinjection working at some point as well.

costinm replied:
Please don't - we have a test that is using manual inject without depending on k8s. 

People look at the tests as examples - we don't want to require access to k8s to do manual injection, it was a mistake.

howardjohn commented on setup/stability-extra/multicluster-vpn/locality-failover/setup.yaml:
```diff
@@ -0,0 +1,86 @@
+apiVersion: v1
```
Failover won't work without an outlierDetection rule like
```
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: outlier
spec:
  host: "google.com"
  trafficPolicy:
    outlierDetection:
      consecutiveErrors: 7
      interval: 5m
      baseEjectionTime: 15m
```

(It can be whatever, just outlierDetection must exist)

howardjohn commented on setup/stability-extra/multicluster-vpn/locality-failover/setup.sh:
```diff
@@ -0,0 +1,71 @@
+    - from: "$REGION2"
+      to: "$REGION1"
+pilot:
+  configNamespace: ${NAMESPACE}
```
You can add `env: PILOT_ENABLE_LOCALITY_LOAD_BALANCING=1` instead of the patch

howardjohn commented on setup/stability-extra/multicluster-vpn/locality-failover/setup.yaml:
```diff
@@ -0,0 +1,86 @@
+        - 10
+        - -qps
+        - 100
+        - http://srv-locality-failover:8080/debug
```
nit: Don't really need the `/debug`

philrud replied:
I want my requests to have some body, but I agree something like "/echo?size=N" is cleaner.

howardjohn commented on setup/stability-extra/multicluster-vpn/setup.sh:
```diff
@@ -0,0 +1,179 @@
+
+# KUBECONFIG: path to a kubeconfig file
+KUBECONFIG1="${KUBECONFIG1:-${HOME}/.kube/config}"
+KUBECONFIG2="${KUBECONFIG2:-${HOME}/.kube/config}"
```
Won't this default to them both being the same cluster? maybe only the first should default?

howardjohn replied:
I see you have the check later so I guess it doesn't matter

philrud replied:
I find my self using either of the two options most often: 1) default kubeconfig, different context names; 2) same context name, two different kubeconfig files (in case I'm generating one on the fly)

So having such defaults allows to pass less arguments on average. And that check below protects the user from the most basic errors such as not passing either of the arguments, etc.

howardjohn replied:
Oh never cool, I didn't think about that

philrud commented:
Addressed most of @howardjohn's comments and rebased.
Not sure how to interpret linter error, doesn't seem to be related to my commit:
```
ERRO Running error: context loading failed: failed to load program with go/packages: go [list -e -json -compiled -test=true -export=false -deps=true -find=false -- ./...]: exit status 1: go: finding github.com/gregjones/httpcache v0.0.0-20181110185634-c63ab54fda8f
```

costinm commented on setup/stability-extra/multicluster-vpn/default/setup.sh:
```diff
@@ -0,0 +1,60 @@
+  configNamespace: ${NAMESPACE}
+EOF
+
+helm template \
```
I would prefer if the iop script from installer was used ( which will be replaced at some point with istioctl ) - or just use kustomize without helm.
